File: SooBin/231211_week/Baekjoon_1753_최단경로.cpp
#include <iostream>
#include <vector>
#include <queue>

#define INF 999999

using namespace std;

// V=정점의 개수, E=간선의 개수, K=시작 정점의 번호
int V, E, K;
// 각 노드에 연결되어 있는 노드에 대한 정보를 담는 배열
vector<pair<int, int>> graph[20001];
// 방문한 적이 있는지 체크하는 목적의 배열
bool visited[20001];
// 최단 거리 테이블
int d[20001];

// 다익스트라 알고리즘
void dijkstra(int start) {
    priority_queue<pair<int, int>> pq;
    // 시작 노드로 가기 위한 최단 경로는 0으로 설정하여 큐에 삽입
    pq.push({0, start});
    d[start] = 0;
    while (!pq.empty()) {
        int dist = -pq.top().first; // 현재 노드까지의 비용
        int now = pq.top().second;  // 현재 노드
        pq.pop();
        // 현재 노드가 이미 처리된 적이 있는 노드라면 무시
        if (d[now] < dist) continue;
        // 현재 노드와 연결된 다른 인접한 노드들을 확인
        for (int i = 0; i < graph[now].size(); i++) {
            // 여기서 second는 큐의 second가 아닌 graph의 second이다.
            // gragh는 출발노드, 도착노드, 비용 순
            int cost = dist + graph[now][i].second;
            // 현재 노드를 거쳐서, 다른 노드로 이동하는 거리가 더 짧은 경우 갱신, 큐에 삽입
            if (cost < d[graph[now][i].first]) {
                d[graph[now][i].first] = cost;
                pq.push(make_pair(-cost, graph[now][i].first));
            }
        }
    }
}


int main() {
 
    cin >> V >> E >> K;

    for (int i = 0; i < E; i++) {
        int a, b, c;
        cin >> a >> b >> c;
        // a노드에서 b노드로 가는 비용은 c이다.
        graph[a].push_back({ b,c });
    }

    // 최단 거리 테이블을 모두 무한으로 초기화
    fill(d, d + 20001, INF);
    dijkstra(K);

    for(int i = 1; i <= V; i++){
        if (d[i] == INF) {
            cout << "INF\n";
        }
        else {
            cout << d[i] << '\n';
        }
    }


    return 0;
}

File: SooYeon/240205_week/Baekjoon_1753_최단경로.js
let fs = require('fs');
let input = fs.readFileSync('dev/stdin').toString().split('\n');

function PriorityQueue(comparator) {
  this._comparator = comparator || PriorityQueue.DEFAULT_COMPARATOR;
  this._elements = [];
}

PriorityQueue.DEFAULT_COMPARATOR = function(a, b) {
  if (typeof a === 'number' && typeof b === 'number') {
    return a - b;
  } else {
    a = a.toString();
    b = b.toString();

    if (a == b) return 0;

    return (a > b) ? 1 : -1;
  }
};

PriorityQueue.prototype.isEmpty = function() {
  return this.size() === 0;
};

PriorityQueue.prototype.peek = function() {
  if (this.isEmpty()) throw new Error('PriorityQueue is empty');

  return this._elements[0];
};

PriorityQueue.prototype.deq = function() {
  var first = this.peek();
  var last = this._elements.pop();
  var size = this.size();

  if (size === 0) return first;

  this._elements[0] = last;
  var current = 0;

  while (current < size) {
    var largest = current;
    var left = (2 * current) + 1;
    var right = (2 * current) + 2;

    if (left < size && this._compare(left, largest) >= 0) {
      largest = left;
    }

    if (right < size && this._compare(right, largest) >= 0) {
      largest = right;
    }

    if (largest === current) break;

    this._swap(largest, current);
    current = largest;
  }

  return first;
};

PriorityQueue.prototype.enq = function(element) {
  var size = this._elements.push(element);
  var current = size - 1;

  while (current > 0) {
    var parent = Math.floor((current - 1) / 2);

    if (this._compare(current, parent) <= 0) break;

    this._swap(parent, current);
    current = parent;
  }

  return size;
};

PriorityQueue.prototype.size = function() {
  return this._elements.length;
};

PriorityQueue.prototype.forEach = function(fn) {
  return this._elements.forEach(fn);
};

PriorityQueue.prototype._compare = function(a, b) {
  return this._comparator(this._elements[a], this._elements[b]);
};

PriorityQueue.prototype._swap = function(a, b) {
  var aux = this._elements[a];
  this._elements[a] = this._elements[b];
  this._elements[b] = aux;
};


// let input = `5 6
// 1
// 5 1 1
// 1 2 2
// 1 3 3
// 2 3 4
// 2 4 5
// 3 4 6`.split('\n');

let [v, e] = input[0].split(' ').map(Number);
let start = Number(input[1]);

let graph = Array.from({length: v + 1}, () => []);
let INF = 987654321;
let distance = Array(v + 1).fill(INF);

for(let i = 2; i <= e + 1; i++) {
  let [a, b, c] = input[i].split(' ').map(Number);
  graph[a].push([b, c]);
}

const dijkstra = (start) => {
  let pq = new PriorityQueue((a, b) => b[0] - a[0]); // 최소힙

  pq.enq([0, start]);
  distance[start] = 0;

  while(pq.size() != 0) {
    let [dist, now] = pq.deq();
    if(distance[now] < dist) continue;

    for(let node of graph[now]) {
      let cost = dist + node[1];

      if(cost < distance[node[0]]) {
        distance[node[0]] = cost;
        pq.enq([cost, node[0]]);
      }
    }
  }
}

dijkstra(start);

let result = distance.slice(1);

let answer = [];
result.forEach((dist) => {
  if(dist === INF) answer.push('INF');
  else answer.push(dist);
})

console.log(answer.join('\n'));